### PR TITLE
refactor(batch): refactor batch `string_agg`

### PR DESCRIPTION
**Pull request by richardchien**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

- Split unordered and ordered implementation.
- Get rid of `DescOrderedRow`, using `OrderedRow` instead.
- Sort the values to aggregate on output, instead of maintaining the order during applying.
- Access `delim` argument via column index instead of `Expression`.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

**Review by BugenZhao — APPROVED** [review]
LGTM

File: src/expr/src/vector_op/agg/string_agg.rs [hunk 1]
```diff
@@ -12,151 +12,181 @@
...
+    ) -> Result<()> {
+        if let (ArrayImpl::Utf8(agg_col), ArrayImpl::Utf8(delim_col)) = (
+            input.column_at(self.agg_col_idx).array_ref(),
+            input.column_at(self.delim_col_idx).array_ref(),
+        ) {
+            for (value, delim) in agg_col
+                .iter()
+                .zip_eq(delim_col.iter())
+                .skip(start_row_id)
+                .take(end_row_id - start_row_id)
```
[the remark this answers is not in the document]
Comment (richardchien) [reply src/expr/src/vector_op/agg/string_agg.rs:92]: I [implemented `nth` before](https://github.com/singularity-data/risingwave/blob/33f26dbd892f4088beba5f38bbe34e59ba3c874a/src/common/src/array/iterator.rs#L79-L87), so it's almost the same

File: src/expr/src/vector_op/agg/string_agg.rs [hunk 2]
```diff
@@ -12,151 +12,181 @@
...
+    }
+
+    fn get_result_and_reset(&mut self) -> Option<String> {
+        let mut rows = std::mem::take(&mut self.unordered_values);
+        if rows.is_empty() {
+            return None;
         }
+        rows.sort_unstable_by(|a, b| a.0.cmp(&b.0));
+        let mut result = rows[0].1.value.clone();
+        for (_, data) in rows.into_iter().skip(1) {
```
[the remark this answers is not in the document]
Comment (richardchien) [reply src/expr/src/vector_op/agg/string_agg.rs]: ok

File: src/expr/src/vector_op/agg/string_agg.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (BugenZhao) [reply src/expr/src/vector_op/agg/string_agg.rs:92]: Cool!

**Pull request merged**